daemon/api.go

```diff
@@ -585,21 +585,50 @@ func deleteOp(c *Command, r *http.Request) Response {
 	}
 }
 
+type licT struct {
+	Intro   string
+	License string
+	Agreed  bool
+}
+
+func (*licT) Error() string {
+	// licT isn't an error, strictly speaking, but can
+	// be converted into one for convenience
+	return "license agreement required"
+}
+
 type packageInstruction struct {
+	progress.NullProgress
 	Action   string `json:"action"`
 	LeaveOld bool   `json:"leave_old"`
+	License  *licT  `json:"license"`
 	pkg      string
-	prog     progress.Meter
 }
 
+func (inst *packageInstruction) Agreed(intro, license string) bool {
+	if inst.License == nil || !inst.License.Agreed || inst.License.Intro != intro || inst.License.License != license {
+		inst.License = &licT{Intro: intro, License: license, Agreed: false}
+		return false
+	}
+
+	return true
+}
+
+var snappyInstall = snappy.Install
+
 func (inst *packageInstruction) install() interface{} {
 	flags := snappy.DoInstallGC
 	if inst.LeaveOld {
 		flags = 0
 	}
-	_, err := snappy.Install(inst.pkg, flags, inst.prog)
-
+	_, err := snappyInstall(inst.pkg, flags, inst)
 	if err != nil {
+		// to be more rigiorous we could also check
+		// err, ok := err.(*snappy.ErrInstallFailed); ok && err.OrigErr == snappy.ErrLicenseNotAccepted
+		// but that seems a bit overkill
```

Comment:
We don't know for sure what err is here, since we're not snappy.Install and we haven't looked at the error. We're basing the heuristics here on our knowledge of how it currently is structured (again, it _is correct_ today, because we know for sure how snappy.Install looks like). But, we don't know how it will look like six months from now.. so the actual right thing to do here is to base behavior on the information we do have: err.

But, again, if you feel strongly about having the heuristics, please just drop the comment that makes it sound like this is a dubious strategy. Either it is fine, or we need to fix it.